### DISCUSSION TITLE
Proposal: Add pragma directive to temporarily promote specific compiler/analyzer warnings to errors

### DISCUSSION THREAD

**Discussion by sindrijo** [dotnet/csharplang#8036] [Language Ideas]
In some contexts it would be useful to be able to force a compiler warning to be an error with a pragma directive. As far as I am aware it's currently only possible to do this by editing the `.editorconfig` file:

```
# CS8509: The switch expression does not handle all possible values of its input type (it is not exhaustive).
dotnet_diagnostic.CS8509.severity = error
```
However that might not always be desirable, so it would be useful to be able to selectively upgrade a warning to an error temporarily with a pragma directive.

```csharp
#pragma warning error CS8509 
```

Example: Unity-3d packages are usually distributed as source code. In this example a package defines an enum type which must be equivalent to another enum type in one of it's dependencies exactly. If the dependency is updated to a version which adds a new entry to the other enum type it should result in a compilation error in this package.

```csharp
internal static SomeOtherPackage.MyEnum ToExternal(this ThisPackage.MyEnum)
{
    // By default a non-exhaustive switch statement is reported as a warning, but in this case we want to explicitly make it an error
    // because ThisPackage would not be fully compatible with the dependency in terms of the enum.

    # pragma warning disable CS8524 // The switch expression does not handle some values of its input type (it is not exhaustive) involving an unnamed enum value.
    # pragma warning error CS8509 // The switch expression does not handle all possible values of its input type (it is not exhaustive).
    return contentEnvironment switch
    {
        SomeOtherPackage.MyEnum.A => ThisPackage.MyEnum.A,
        SomeOtherPackage.MyEnum.B => ThisPackage.MyEnum.B,
        // If SomeOtherPackage adds MyEnum.C this switch statement should induce error because it is no longer exhaustive
    };
    # pragma warning restore CS8509
    # pragma warning restore CS8524
}
```

NB this feature already exists in CPP: https://learn.microsoft.com/en-us/cpp/preprocessor/warning?view=msvc-170

**Comment by 333fred** [comment]
Duplicate of https://github.com/dotnet/csharplang/issues/3986.

**Reply by sindrijo** [reply]
This doesn't seem like it's a duplicate. This is different, this is about temporarily elevating a warning to an error, found nothing in the linked items that relates to this.

**Reply by 333fred** [reply]
That is exactly what the issue I linked is: elevating a diagnostic to an error. We explicitly rejected that due to the adverse effects of will have on the IDE and the relative nicheness of the need.
